File: obs-nvidia-notify/shared/src/protocol.cpp
// SPDX-License-Identifier: MIT
#include "obsn/protocol.hpp"

#include <algorithm>

namespace obsn::proto {
namespace {

struct TypeEntry {
    const char* name;
    MessageType type;
};

constexpr TypeEntry kTypes[] = {
    {"hello", MessageType::Hello},
    {"state_snapshot", MessageType::StateSnapshot},
    {"event", MessageType::Event},
    {"heartbeat", MessageType::Heartbeat},
    {"error", MessageType::Error},
    {"client_hello", MessageType::ClientHello},
    {"request_snapshot", MessageType::RequestSnapshot},
    {"ping", MessageType::Ping},
    {"pong", MessageType::Pong},
};

}  // namespace

const char* to_string(MessageType type) noexcept {
    for (const TypeEntry& e : kTypes) {
        if (e.type == type) return e.name;
    }
    return "unknown";
}

MessageType parse_message_type(std::string_view text) noexcept {
    for (const TypeEntry& e : kTypes) {
        if (text == e.name) return e.type;
    }
    return MessageType::Unknown;
}

bool is_producer_to_consumer(MessageType type) noexcept {
    switch (type) {
        case MessageType::Hello:
        case MessageType::StateSnapshot:
        case MessageType::Event:
        case MessageType::Heartbeat:
        case MessageType::Error:
        case MessageType::Pong:
            return true;
        case MessageType::ClientHello:
        case MessageType::RequestSnapshot:
        case MessageType::Ping:
        case MessageType::Unknown:
            return false;
    }
    return false;
}

std::string Envelope::encode() const {
    json::Object o;
    o.emplace_back("v", json::Value(version));
    o.emplace_back("seq", json::Value(seq));
    o.emplace_back("ts", json::Value(ts));
    o.emplace_back("type", json::Value(to_string(type)));
    if (!data.is_null()) o.emplace_back("data", data);
    return json::Value(std::move(o)).dump();
}

DecodeResult decode(std::string_view line) {
    DecodeResult result;
    if (line.size() > kMaxMessageBytes) {
        result.status = DecodeStatus::TooLarge;
        result.error = "message exceeds the " + std::to_string(kMaxMessageBytes) + " byte limit";
        return result;
    }

    json::Limits limits;
    limits.max_total_bytes = kMaxMessageBytes;
    const json::ParseResult parsed = json::parse(line, limits);
    if (!parsed.ok) {
        result.status = DecodeStatus::NotJson;
        // The parser's error never quotes payload content, so this is safe to log verbatim.
        result.error = parsed.error;
        return result;
    }
    if (!parsed.value.is_object()) {
        result.status = DecodeStatus::NotObject;
        result.error = "top-level value is not an object";
        return result;
    }

    const json::Value& doc = parsed.value;
    if (!doc.has("type")) {
        result.status = DecodeStatus::MissingField;
        result.error = "missing 'type'";
        return result;
    }

    Envelope envelope;
    envelope.version = static_cast<int>(doc.get_int("v", kProtocolVersion));
    // The version is checked before the type, because a peer speaking a protocol we do not know
    // may well be using type names that mean something else entirely.
    if (envelope.version < kProtocolMin || envelope.version > kProtocolMax) {
        result.status = DecodeStatus::BadVersion;
        result.error = "protocol version " + std::to_string(envelope.version) +
                       " is outside the supported range " + std::to_string(kProtocolMin) + "-" +
                       std::to_string(kProtocolMax);
        return result;
    }

    envelope.seq = doc.get_int("seq");
    envelope.ts = doc.get_int("ts");
    envelope.type = parse_message_type(doc.get_string("type"));
    if (envelope.type == MessageType::Unknown) {
        result.status = DecodeStatus::UnknownType;
        // Not fatal: a newer producer sending a type this build does not know is a message to
        // skip, not a reason to drop a working connection.
        result.error = "unrecognised message type";
        result.envelope = std::move(envelope);
        return result;
    }
    if (const json::Value* data = doc.find("data")) envelope.data = *data;

    result.status = DecodeStatus::Ok;
    result.envelope = std::move(envelope);
    return result;
}

json::Value HelloPayload::to_json() const {
    json::Object o;
    o.emplace_back("protocol_min", json::Value(protocol_min));
    o.emplace_back("protocol_max", json::Value(protocol_max));
    o.emplace_back("script_version", json::Value(script_version));
    o.emplace_back("obs_version", json::Value(obs_version));
    o.emplace_back("platform", json::Value(platform));
    json::Array caps;
    caps.reserve(capabilities.size());
    for (const std::string& c : capabilities) caps.emplace_back(c);
    o.emplace_back("capabilities", json::Value(std::move(caps)));
    return json::Value(std::move(o));
}

HelloPayload HelloPayload::from_json(const json::Value& v) {
    HelloPayload p;
    p.protocol_min = static_cast<int>(v.get_int("protocol_min", kProtocolMin));
    p.protocol_max = static_cast<int>(v.get_int("protocol_max", kProtocolMax));
    p.script_version = v.get_string("script_version");
    p.obs_version = v.get_string("obs_version");
    p.platform = v.get_string("platform");
    if (const json::Value* caps = v.find("capabilities"); caps != nullptr && caps->is_array()) {
        for (const json::Value& c : caps->as_array()) {
            if (c.is_string()) p.capabilities.push_back(c.as_string());
        }
    }
    return p;
}

json::Value ClientHelloPayload::to_json() const {
    json::Object o;
    o.emplace_back("protocol", json::Value(protocol));
    o.emplace_back("client", json::Value(client));
    o.emplace_back("client_version", json::Value(client_version));
    o.emplace_back("process", json::Value(process));
    o.emplace_back("pid", json::Value(pid));
    return json::Value(std::move(o));
}

ClientHelloPayload ClientHelloPayload::from_json(const json::Value& v) {
    ClientHelloPayload p;
    p.protocol = static_cast<int>(v.get_int("protocol", kProtocolVersion));
    p.client = v.get_string("client", "unknown");
    p.client_version = v.get_string("client_version");
    p.process = v.get_string("process");
    p.pid = v.get_int("pid");
    return p;
}

json::Value ErrorPayload::to_json() const {
    json::Object o;
    o.emplace_back("code", json::Value(code));
    o.emplace_back("message", json::Value(message));
    return json::Value(std::move(o));
}

ErrorPayload ErrorPayload::from_json(const json::Value& v) {
    ErrorPayload p;
    p.code = v.get_string("code");
    p.message = v.get_string("message");
    return p;
}

Envelope make(MessageType type, json::Value data, std::int64_t seq, std::int64_t ts) {
    Envelope e;
    e.type = type;
    e.data = std::move(data);
    e.seq = seq;
    e.ts = ts;
    return e;
}

void clamp_event(ObsEvent& event) {
    event.detail = json::truncate_utf8(event.detail, kMaxDetailChars);
    event.path = json::truncate_utf8(event.path, kMaxPathChars);
    event.scene = json::truncate_utf8(event.scene, kMaxDetailChars);
    event.previous_scene = json::truncate_utf8(event.previous_scene, kMaxDetailChars);
    event.profile = json::truncate_utf8(event.profile, kMaxDetailChars);
    event.service = json::truncate_utf8(event.service, kMaxDetailChars);
    // Negative durations and sizes are the "not known" sentinel; anything else negative is a
    // producer bug and is normalised to that same sentinel rather than formatted.
    if (event.duration_ms < 0) event.duration_ms = -1;
    if (event.size_bytes < 0) event.size_bytes = -1;
    if (event.attempt < 0) event.attempt = 0;
    if (event.replay_seconds < 0) event.replay_seconds = 0;
}

}  // namespace obsn::proto

File: obs-nvidia-notify/tools/obsn-cli/src/protocol.ts
// SPDX-License-Identifier: MIT
// A second, independent implementation of the wire format.
//
// This exists to be checked against the C++ one rather than to share code with it: a format
// that two implementations agree on is a format that is actually specified, and the parity
// tests are what turn docs/protocol.md from a description into a contract.

export const PROTOCOL_VERSION = 1;
export const MAX_MESSAGE_BYTES = 65536;

export type MessageType =
  | "hello"
  | "state_snapshot"
  | "event"
  | "heartbeat"
  | "error"
  | "client_hello"
  | "request_snapshot"
  | "ping"
  | "pong";

/** Types that may legitimately travel producer to consumer. */
export const PRODUCER_TO_CONSUMER: ReadonlySet<MessageType> = new Set<MessageType>([
  "hello",
  "state_snapshot",
  "event",
  "heartbeat",
  "error",
  "pong",
]);

export type EventKind =
  | "recording.starting"
  | "recording.started"
  | "recording.stopping"
  | "recording.stopped"
  | "recording.paused"
  | "recording.resumed"
  | "recording.saved"
  | "replay.starting"
  | "replay.started"
  | "replay.stopped"
  | "replay.saved"
  | "stream.starting"
  | "stream.started"
  | "stream.stopping"
  | "stream.stopped"
  | "stream.reconnecting"
  | "stream.reconnected"
  | "virtualcam.started"
  | "virtualcam.stopped"
  | "scene.changed"
  | "profile.changed"
  | "warning"
  | "script.connected"
  | "script.disconnected";

export const EVENT_KINDS: readonly EventKind[] = [
  "recording.starting", "recording.started", "recording.stopping", "recording.stopped",
  "recording.paused", "recording.resumed", "recording.saved",
  "replay.starting", "replay.started", "replay.stopped", "replay.saved",
  "stream.starting", "stream.started", "stream.stopping", "stream.stopped",
  "stream.reconnecting", "stream.reconnected",
  "virtualcam.started", "virtualcam.stopped",
  "scene.changed", "profile.changed", "warning",
  "script.connected", "script.disconnected",
];

export type OutputState =
  | "idle" | "starting" | "active" | "paused" | "stopping" | "reconnecting";

export interface ObsEvent {
  kind: EventKind;
  ts: number;
  path?: string;
  /** Milliseconds. Absent means "not known" -- never 0, which would be a measurement. */
  duration_ms?: number;
  size_bytes?: number;
  scene?: string;
  previous_scene?: string;
  profile?: string;
  service?: string;
  reason?: string;
  detail?: string;
  attempt?: number;
  replay_seconds?: number;
}

export interface ObsState {
  obs_version?: string;
  script_version?: string;
  profile?: string;
  scene_collection?: string;
  current_scene?: string;
  recording?: { state: OutputState; started_ms?: number; paused_ms?: number;
                paused_since_ms?: number; path?: string };
  replay?: { state: OutputState; started_ms?: number; last_saved_path?: string;
             last_saved_ms?: number; duration_s?: number };
  stream?: { state: OutputState; started_ms?: number; service?: string;
             reconnect_attempt?: number };
  virtual_cam?: { state: OutputState; started_ms?: number };
  stats?: Record<string, number>;
}

export interface Envelope {
  v: number;
  seq: number;
  ts: number;
  type: MessageType;
  data?: unknown;
}

export type DecodeStatus =
  | "ok" | "not_json" | "not_object" | "missing_field" | "bad_version"
  | "unknown_type" | "too_large";

export interface DecodeResult {
  status: DecodeStatus;
  envelope?: Envelope;
  error?: string;
}

/** Whether a decode failure warrants dropping the connection, as opposed to one message. */
export function isFatal(status: DecodeStatus): boolean {
  return status === "not_json" || status === "too_large" || status === "not_object";
}

export function encode(type: MessageType, seq: number, ts: number, data?: unknown): string {
  const envelope: Envelope = { v: PROTOCOL_VERSION, seq, ts, type };
  if (data !== undefined) envelope.data = data;
  return `${JSON.stringify(envelope)}\n`;
}

export function decode(line: string): DecodeResult {
  if (Buffer.byteLength(line, "utf8") > MAX_MESSAGE_BYTES) {
    return { status: "too_large", error: "message exceeds the size limit" };
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(line);
  } catch (error) {
    return { status: "not_json", error: String(error) };
  }
  if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
    return { status: "not_object", error: "top-level value is not an object" };
  }
  const doc = parsed as Record<string, unknown>;
  if (typeof doc["type"] !== "string") {
    return { status: "missing_field", error: "missing 'type'" };
  }
  // The version is checked before the type, because a peer speaking a protocol we do not know
  // may well use type names that mean something else entirely.
  const version = typeof doc["v"] === "number" ? doc["v"] : PROTOCOL_VERSION;
  if (version !== PROTOCOL_VERSION) {
    return { status: "bad_version", error: `unsupported protocol version ${version}` };
  }
  const type = doc["type"] as MessageType;
  if (!isMessageType(type)) {
    return { status: "unknown_type", error: "unrecognised message type" };
  }
  return {
    status: "ok",
    envelope: {
      v: version,
      seq: typeof doc["seq"] === "number" ? doc["seq"] : 0,
      ts: typeof doc["ts"] === "number" ? doc["ts"] : 0,
      type,
      data: doc["data"],
    },
  };
}

const MESSAGE_TYPES: ReadonlySet<string> = new Set([
  "hello", "state_snapshot", "event", "heartbeat", "error",
  "client_hello", "request_snapshot", "ping", "pong",
]);

export function isMessageType(value: string): value is MessageType {
  return MESSAGE_TYPES.has(value);
}

/**
 * Splits a byte stream into lines with a hard cap enforced *during* accumulation.
 *
 * The cap matters: a peer that sends a megabyte without a newline must not be able to grow the
 * buffer to a megabyte before anyone notices.
 */
export class LineFramer {
  #pending = "";
  #overflowed = false;

  constructor(private readonly maxBytes = MAX_MESSAGE_BYTES) {}

  get overflowed(): boolean {
    return this.#overflowed;
  }

  get pendingBytes(): number {
    return Buffer.byteLength(this.#pending, "utf8");
  }

  reset(): void {
    this.#pending = "";
    this.#overflowed = false;
  }

  /** Returns the complete lines, or null once the framer has overflowed. */
  feed(chunk: string): string[] | null {
    if (this.#overflowed) return null;
    this.#pending += chunk;
    const lines: string[] = [];
    let index = this.#pending.indexOf("\n");
    while (index >= 0) {
      lines.push(this.#pending.slice(0, index));
      this.#pending = this.#pending.slice(index + 1);
      index = this.#pending.indexOf("\n");
    }
    if (this.pendingBytes > this.maxBytes) {
      this.#overflowed = true;
      this.#pending = "";
      // Lines already completed are still returned: they were whole and are not the problem.
      return lines.length > 0 ? lines : null;
    }
    return lines;
  }
}
